kubernetes-informer/src/main/java/io/micronaut/kubernetes/client/informer/ResourceEventHandlerConstructorInterceptor.java

```diff
@@ -0,0 +1,177 @@
+import io.kubernetes.client.informer.ResourceEventHandler;
+import io.kubernetes.client.informer.SharedIndexInformer;
+import io.kubernetes.client.informer.SharedInformerFactory;
+import io.kubernetes.client.util.Namespaces;
+import io.micronaut.aop.ConstructorInterceptor;
+import io.micronaut.aop.ConstructorInvocationContext;
+import io.micronaut.aop.InterceptorBean;
+import io.micronaut.context.ApplicationContext;
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.kubernetes.client.DiscoveryCache;
+import io.micronaut.kubernetes.client.NamespaceResolver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+/**
+ * Constructor interceptor for the {@link ResourceEventHandler} beans annotated by {@link Informer} annotation that
+ * based on provided parameters in the {@link Informer} annotation created the {@link SharedIndexInformer} and registers
+ * {@link ResourceEventHandler} to the informer.
+ *
+ * @param <ApiType> type of Kubernetes Object
+ * @author Pavol Gressa
+ * @since 3.1
+ */
+@Requires(beans = SharedInformerFactory.class)
+@InterceptorBean(Informer.class)
+@Internal
+public class ResourceEventHandlerConstructorInterceptor<ApiType extends KubernetesObject> implements ConstructorInterceptor<ResourceEventHandler<ApiType>> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ResourceEventHandlerConstructorInterceptor.class);
+
+    private final SharedIndexInformerFactory sharedIndexInformerFactory;
+    private final NamespaceResolver namespaceResolver;
+    private final ApplicationContext applicationContext;
+    private final DiscoveryCache discoveryCache;
+
+    public ResourceEventHandlerConstructorInterceptor(SharedIndexInformerFactory sharedIndexInformerFactory,
+                                                      NamespaceResolver namespaceResolver,
+                                                      ApplicationContext applicationContext,
+                                                      @Nullable DiscoveryCache discoveryCache) {
+        this.sharedIndexInformerFactory = sharedIndexInformerFactory;
+        this.namespaceResolver = namespaceResolver;
+        this.applicationContext = applicationContext;
+        this.discoveryCache = discoveryCache;
+    }
+
+    @Override
+    public @NonNull
+    ResourceEventHandler<ApiType> intercept(@NonNull ConstructorInvocationContext<ResourceEventHandler<ApiType>> context) {
+        final Class<ResourceEventHandler<ApiType>> declaringType = context.getDeclaringType();
+
+        if (declaringType.isAnnotationPresent(Informer.class)) {
+            Informer typeAnnotation = declaringType.getAnnotation(Informer.class);
```

Comment:
for this to work on Graal you will probably have to define the proxy metadata for the `Informer` annotation